### PR TITLE
Use scikit-build for the build process

### PR DESCRIPTION
**Pull request by vyasr**
This PR replaces a lot of the setuptools logic with using scikit-build, which allows us to leverage the existing CMake configuration for the C++ components of RMM to build the Python package using standard pip- and setup.py-based installs. Replaces #637.

### CODE REVIEW

**Comment by vyasr** [comment]
CC @kkraus14

File: python/CMakeLists.txt
Comment (kkraus14) [inline python/CMakeLists.txt:27]: Do we need the `CUDA` language here? Isn't it all host code?

**Review by kkraus14 — COMMENTED** [review]
One question, otherwise this looks amazing!

File: python/CMakeLists.txt
Comment (kkraus14) [reply python/CMakeLists.txt:27]: I believe we don't even use the `CUDA` language in the RMM C++ project so I can't imagine it being an issue here.

https://github.com/rapidsai/rmm/blob/branch-22.04/CMakeLists.txt#L29

Comment (vyasr) [reply python/CMakeLists.txt:27]: I was able to remove the `find_package` for CUDAToolkit because that is correctly being pulled in from adding the rmm include directory for the C++ build if needed. However, it seems that because we are linking to rmm with `target_link_libraries(${cython_module} rmm::rmm)` we also need to include the languages required by that link. If I remove the CUDA language I get a CMake error `Cannot use features from non-enabled language CUDA`. If I comment out that link line then the configuration succeeds (of course the build fails because it can't find RMM headers).

I find it interesting that the configuration for the C++ part of the project doesn't list CUDA as a language at all though. However, it does link to `CUDA::*` targets as interface libraries. Perhaps that is why the dependency is being transmitted to the Python build even though the C++ config doesn't require CUDA? @robertmaynard could you shed any light on what might be going on here?

Comment (vyasr) [reply python/CMakeLists.txt:27]: Ahh I think I see it. Since RMM is header-only, it doesn't need CUDA *unless you build tests or benchmarks*. And indeed, we see that the language is conditionally enabled here: https://github.com/rapidsai/rmm/blob/branch-22.04/CMakeLists.txt#L84.

Comment (kkraus14) [reply python/CMakeLists.txt:27]: My best guess is it's coming from the build interface definition here: https://github.com/rapidsai/rmm/blob/branch-22.04/CMakeLists.txt#L79

I think that's being set so any CUDA code that includes RMM inherits the C++17 requirement, but in our case here it forces the need for CUDA language unnecessarily.

**Comment by vyasr** [comment]
rerun tests

File: python/CMakeLists.txt
Comment (harrism) [inline python/CMakeLists.txt:29]: C++ for humans. CPP for parsers. :)

```suggestion
option(FIND_RMM_CPP "Search for existing RMM C++ installations before defaulting to local files"
```

**Review by harrism — APPROVED** [review]
Very cool.  One comment suggestion.

**Comment by vyasr** [comment]
Waiting on new build environments to ensure that scikit-build is actually installed on all our CI executors so that we can isolate the real remaining failures. There appear to be some outstanding issues causing seg faults on some platforms, but I can't repro them locally so I'm waiting to get a clean indication from CI of what real issues remain.

File: python/CMakeLists.txt
Comment (harrism) [inline python/CMakeLists.txt:26]: Suggest putting TODOs on their own lines for visibility.

```suggestion
  # keep this. 
  # TODO: The C++ RMM CMake configuration targets cuda_std_17 features, which prior to
```

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
AFAICT all the failing tests are numba test, and I have tried running tests in the ubuntu 18.04 CI container locally and all tests pass, so I am suspecting some sort of version mismatch (perhaps driver/runtime?) between the cudatoolkit in the container and something on the host machine being discovered differently by numba and rmm.

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
rerun tests

File: python/CMakeLists.txt
Comment (vyasr) [reply python/CMakeLists.txt]: cmake-format made this request quite difficult to accommodate, but it's a good call. I think the result makes it much clearer what each TODO is referring to.

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: python/CMakeLists.txt
Comment (kkraus14) [inline python/CMakeLists.txt:73]: By default with a `find_package` call a user can provide an `RMM_ROOT` CMake or environment variable to influence where it searches to find the package. If I remember correctly we did the work to make sure that an rmm build directory could suffice for this call, but I could be mistaken.

**Review by kkraus14 — APPROVED** [review]
One minor comment, otherwise LGTM. Thanks @vyasr

File: python/CMakeLists.txt
Comment (vyasr) [reply python/CMakeLists.txt:73]: When you say that we did the work, are you referring to your original scikit-build PR or the current RMM CMake setting the appropriate environment variables to make itself discoverable?

**Comment by vyasr** [comment]
I'm blocking merging for now to make sure that we update anything downstream of RMM that needs to account for the changes to the build. Also, we're just over a week away from the start of burn down. While we could merge this now, I'm a little bit antsy about that and it may be better to wait until after the release so that we have a full release cycle of devs testing this before we put it into a release (and we can start doing this for other RAPIDS packages as well). Thoughts @robertmaynard @ajschmidt8 @harrism @JohnZed?

**Comment by ajschmidt8** [comment]
> I'm blocking merging for now to make sure that we update anything downstream of RMM that needs to account for the changes to the build. Also, we're just over a week away from the start of burn down. While we could merge this now, I'm a little bit antsy about that and it may be better to wait until after the release so that we have a full release cycle of devs testing this before we put it into a release (and we can start doing this for other RAPIDS packages as well). Thoughts @robertmaynard @ajschmidt8 @harrism @JohnZed?

No qualms here. If this PR is not urgent, I'm fine with waiting.

File: python/CMakeLists.txt
Comment (robertmaynard) [reply python/CMakeLists.txt:73]: current librmm can be found via the build directory. librmm doesn't need to do anything, the call to rmm would specify one of the following:

- `RMM_ROOT` ( env var or CMake var )
- `CMAKE_PREFIX_PATH` ( env var or CMake var )

**Comment by vyasr** [comment]
I have verified that tests pass for both cuML and cuSpatial with this build of RMM locally, so that's a good sign. I'm going to meet with @ajschmidt8 to test some conda package builds using this branch in a fresh environment. If that works fine, then I think we may as well push out the changes and see how it goes.

**Comment by harrism** [comment]
That's great, thanks for the comprehensive testing! This will be a great improvement to the library.

**Comment by kkraus14** [comment]
Got a chance to test this and the only thing I see is that a `rmm/_lib/tests/__init__.py` file is being included in the package:
```
python setup.py bdist_wheel
```
```
(dev) keith@Keith-PC:~/git/rmm/python/dist$ python -m zipfile --list rmm-22.4.0a0+78.g4867f970-cp38-cp38-linux_x86_64.whl
File Name                                             Modified             Size
rmm/__init__.py                                2022-03-08 16:57:38          973
rmm/_version.py                                2022-03-08 16:57:38          514
rmm/mr.py                                      2022-03-08 16:57:38         1943
rmm/rmm.py                                     2022-03-08 16:57:38         7795
rmm/_cuda/__init__.pxd                         2022-03-08 16:57:38            0
rmm/_cuda/__init__.py                          2022-03-08 16:57:38            0
rmm/_cuda/gpu.pxd                              2022-03-08 16:57:38        16217
rmm/_cuda/gpu.py                               2022-03-08 16:57:38         4366
rmm/_cuda/stream.cpython-38-x86_64-linux-gnu.so 2022-03-08 16:57:36       401096
rmm/_cuda/stream.pxd                           2022-03-08 16:57:38         1178
rmm/_lib/__init__.pxd                          2022-03-08 16:57:38          593
rmm/_lib/__init__.py                           2022-03-08 16:57:38          634
rmm/_lib/cuda_stream.cpython-38-x86_64-linux-gnu.so 2022-03-08 16:57:36       373824
rmm/_lib/cuda_stream.pxd                       2022-03-08 16:57:38         1244
rmm/_lib/cuda_stream_pool.pxd                  2022-03-08 16:57:38          941
rmm/_lib/cuda_stream_view.pxd                  2022-03-08 16:57:38         1170
rmm/_lib/device_buffer.cpython-38-x86_64-linux-gnu.so 2022-03-08 16:57:38       679096
rmm/_lib/device_buffer.pxd                     2022-03-08 16:57:38         3059
rmm/_lib/device_uvector.pxd                    2022-03-08 16:57:38         1571
rmm/_lib/lib.cpython-38-x86_64-linux-gnu.so    2022-03-08 16:57:36        23032
rmm/_lib/lib.pxd                               2022-03-08 16:57:38          775
rmm/_lib/memory_resource.cpython-38-x86_64-linux-gnu.so 2022-03-08 16:57:36       788352
rmm/_lib/memory_resource.pxd                   2022-03-08 16:57:38         2156
rmm/_lib/tests/__init__.py                     2022-03-08 16:57:38            0
...
```

Probably not an issue in practice.

**Comment by vyasr** [comment]
@ajschmidt8 and I just met and went through installing these conda packages into a fresh environment. Everything looks good, the package is well-formed and Python tests pass so at this point I think I'm comfortable proceeding. I think it's very unlikely that we will run into conda package issues at this point, and if someone finds bugs trying to build from source due to unexpected system configuration I expect them to be a small enough portion of the user base that we can work to resolve those issues if and when they arise.

**Comment by vyasr** [comment]
@robertmaynard @rapidsai/rmm-python-codeowners please take a final look at this PR, we can merge once you've approved.

File: build.sh
Comment (robertmaynard) [inline build.sh:40]:
```suggestion
RMM_BUILD_DIRS="${REPODIR}/python/build ${REPODIR}/python/_skbuild"
```

File: build.sh
Comment (robertmaynard) [inline build.sh:183]: Indentation is off here

File: python/CMakeLists.txt
Comment (robertmaynard) [inline python/CMakeLists.txt:18]: We should use `update-version.sh` to keep the version string in check.

What you want is:

```cmake
project(
  rmm-python
  VERSION 22.04.00
  ...
  )
  
You will also need to update the  `update-version.sh` to modify the `VERSION` entry in this file 
```

File: python/CMakeLists.txt
Comment (robertmaynard) [inline python/CMakeLists.txt:43]: I would prefer we move this out of the CMakeLists.txt and into a support file

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: build.sh
Comment (vyasr) [reply build.sh:183]: Tabs snuck in, thanks.

**Review by robertmaynard — APPROVED** [review]

File: build.sh
Comment (galipremsagar) [inline build.sh:160]:
```suggestion
    fi
```
Seems like a misaligned `fi`?

File: build.sh
Comment (galipremsagar) [inline build.sh:183]:
```suggestion
        python setup.py build_ext --inplace -DCUDA_MALLOC_ASYNC_SUPPORT=OFF
```

File: python/dev_requirements.txt
Comment (galipremsagar) [inline python/dev_requirements.txt:1]:
```suggestion
# Copyright (c) 2021-2022, NVIDIA CORPORATION.
```

File: build.sh
Comment (vyasr) [reply build.sh:160]: Ack thanks. In removing tabs I accidentally deindented this.

File: build.sh
Comment (vyasr) [reply build.sh:183]: Actually you need the extra `--`. In general, bash scripts treat `--` as a way to terminate a list of options. Invoking a setup.py script using scikit-build consists of three separate layers of execution: 1) the setup.py script itself, 2) the CMake command invocations, and 3) the underlying build tool (e.g. Ninja or make) invocation. You can specify options to each of these three steps using a suitable number of separators:
```
python setup.py install --SETUPTOOLS_OPTIONS -- --CMAKE_OPTION -- --NINJA_OPTION
```
In this case, we're passing a CMake option, so we need one `--`.

**Review by galipremsagar — APPROVED** [review]

File: CONTRIBUTING.md
Comment (kkraus14) [inline CONTRIBUTING.md:37]:
```suggestion
To set up a development environment, follow the steps in the [README](https://github.com/rapidsai/rmm/blob/main/README.md) for cloning the repository and creating the conda environment.
```

File: build.sh
Comment (kkraus14) [inline build.sh:185]: I think this needs the `--` as well?
```suggestion
        python setup.py build_ext --inplace -- -DCMAKE_PREFIX_PATH=${INSTALL_PREFIX}
```

Comment (vyasr) [reply build.sh:185]: Thanks for catching that

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
